File: src/lib/import/service.rs
use crate::import::model::{AnchoredSubtitleRow, CsvSubtitleRow};
use crate::import::ports::ImportService;
use crate::subtitle::model::SubtitleError;
use csv::ReaderBuilder;
use log::debug;
use std::fs;
use std::path::Path;

pub struct CsvImportService;

impl CsvImportService {
    pub fn new() -> Self {
        Self
    }

    fn validate_csv_header(
        headers: &csv::StringRecord,
        delimiter: char,
    ) -> Result<(), SubtitleError> {
        let actual = headers
            .iter()
            .collect::<Vec<_>>()
            .join(&delimiter.to_string());

        if headers.len() != 3
            || headers.get(0) != Some("start_time")
            || headers.get(1) != Some("end_time")
            || headers.get(2) != Some("text")
        {
            return Err(SubtitleError::InvalidCsvHeader(
                delimiter.to_string(),
                actual,
            ));
        }

        debug!("csv header validated successfully");
        Ok(())
    }
}

impl ImportService for CsvImportService {
    fn parse_csv_file(
        &self,
        csv_path: &Path,
        delimiter: char,
    ) -> Result<Vec<CsvSubtitleRow>, SubtitleError> {
        debug!(
            "parsing csv file: {:?} with delimiter '{}'",
            csv_path, delimiter
        );

        let mut reader = ReaderBuilder::new()
            .delimiter(delimiter as u8)
            .has_headers(true)
            .from_path(csv_path)
            .map_err(|e| SubtitleError::IoError(e.into()))?;

        let headers = reader.headers().map_err(|e| SubtitleError::CsvParseError {
            line: 1,
            message: format!("failed to read CSV header: {}", e),
        })?;

        debug!("validating csv header");
        Self::validate_csv_header(headers, delimiter)?;

        let mut rows = Vec::new();

        for (idx, result) in reader.records().enumerate() {
            let line_number = idx + 2; // +1 for 0-index, +1 for header

            let record = result.map_err(|e| SubtitleError::CsvParseError {
                line: line_number,
                message: format!("failed to parse CSV record: {}", e),
            })?;

            if record.len() != 3 {
                return Err(SubtitleError::CsvParseError {
                    line: line_number,
                    message: format!("expected 3 columns, got {}", record.len()),
                });
            }

            debug!(
                "parsed csv row {}: start={}, end={}",
                line_number, &record[0], &record[1]
            );

            rows.push(CsvSubtitleRow {
                line_number,
                start_time: record[0].trim().to_string(),
                end_time: record[1].trim().to_string(),
                text: record[2].replace("\\n", "\n"), // Convert \n to actual newlines
            });
        }

        debug!("parsed {} rows from csv file", rows.len());
        Ok(rows)
    }

    fn parse_anchored_file(
        &self,
        _anchored_path: &Path,
    ) -> Result<Vec<AnchoredSubtitleRow>, SubtitleError> {
        Err(SubtitleError::InvalidPath(
            "CsvImportService does not support anchored format parsing".to_string(),
        ))
    }
}

pub struct AnchoredImportService;

impl AnchoredImportService {
    pub fn new() -> Self {
        Self
    }

    /// Parse anchored format line to extract index and first line text
    fn parse_anchored_line(line: &str) -> Option<(u32, String)> {
        let trimmed = line.trim();
        if !trimmed.starts_with('[') {
            return None;
        }

        let close_bracket = trimmed.find(']')?;
        let index_str = &trimmed[1..close_bracket];
        let index = index_str.parse::<u32>().ok()?;

        let text = if close_bracket + 1 < trimmed.len() {
            trimmed[close_bracket + 1..].trim_start().to_string()
        } else {
            String::new()
        };

        Some((index, text))
    }
}

impl ImportService for AnchoredImportService {
    fn parse_csv_file(
        &self,
        _csv_path: &Path,
        _delimiter: char,
    ) -> Result<Vec<CsvSubtitleRow>, SubtitleError> {
        Err(SubtitleError::InvalidPath(
            "AnchoredImportService does not support CSV parsing".to_string(),
        ))
    }

    fn parse_anchored_file(
        &self,
        anchored_path: &Path,
    ) -> Result<Vec<AnchoredSubtitleRow>, SubtitleError> {
        debug!("parsing anchored format file: {:?}", anchored_path);

        let content = fs::read_to_string(anchored_path)?;

        let mut rows = Vec::new();
        let mut current_index: Option<u32> = None;
        let mut current_text = String::new();
        let mut current_line_number = 0;

        for (line_num, line) in content.lines().enumerate() {
            let line_number = line_num + 1;

            if let Some((index, first_line_text)) = Self::parse_anchored_line(line) {
                if let Some(idx) = current_index {
                    if current_text.trim().is_empty() {
                        return Err(SubtitleError::AnchoredParseError {
                            line: current_line_number,
                            message: format!("Subtitle [{}] has empty text", idx),
                        });
                    }
                    rows.push(AnchoredSubtitleRow {
                        line_number: current_line_number,
                        index: idx,
                        text: current_text.trim().to_string(),
                    });
                }

                current_index = Some(index);
                current_text = first_line_text;
                current_line_number = line_number;
            } else {
                if current_index.is_some() {
                    if !current_text.is_empty() {
                        current_text.push('\n');
                    }
                    current_text.push_str(line);
                } else if !line.trim().is_empty() {
                    return Err(SubtitleError::AnchoredParseError {
                        line: line_number,
                        message: "Text line found before any [INDEX] marker".to_string(),
                    });
                }
            }
        }

        if let Some(idx) = current_index {
            if current_text.trim().is_empty() {
                return Err(SubtitleError::AnchoredParseError {
                    line: current_line_number,
                    message: format!("Subtitle [{}] has empty text", idx),
                });
            }
            rows.push(AnchoredSubtitleRow {
                line_number: current_line_number,
                index: idx,
                text: current_text.trim().to_string(),
            });
        }

        if rows.is_empty() {
            return Err(SubtitleError::AnchoredParseError {
                line: 0,
                message: "No valid subtitle entries found in anchored file".to_string(),
            });
        }

        debug!("parsed {} entries from anchored file", rows.len());
        Ok(rows)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;
    use tempfile::NamedTempFile;

    #[test]
    fn test_parse_csv_valid() {
        let mut temp_file = NamedTempFile::new().unwrap();
        writeln!(temp_file, "start_time|end_time|text").unwrap();
        writeln!(temp_file, "00:00:01,000|00:00:02,000|First subtitle").unwrap();
        writeln!(temp_file, "00:00:03,000|00:00:04,000|Second subtitle").unwrap();
        temp_file.flush().unwrap();

        let service = CsvImportService::new();
        let rows = service.parse_csv_file(temp_file.path(), '|').unwrap();
        assert_eq!(rows.len(), 2);
        assert_eq!(rows[0].line_number, 2);
        assert_eq!(rows[0].start_time, "00:00:01,000");
        assert_eq!(rows[0].end_time, "00:00:02,000");
        assert_eq!(rows[0].text, "First subtitle");
        assert_eq!(rows[1].line_number, 3);
        assert_eq!(rows[1].text, "Second subtitle");
    }

    #[test]
    fn test_parse_csv_multiline_text() {
        let mut temp_file = NamedTempFile::new().unwrap();
        writeln!(temp_file, "start_time|end_time|text").unwrap();
        writeln!(temp_file, "00:00:01,000|00:00:02,000|Line 1\\nLine 2").unwrap();
        temp_file.flush().unwrap();

        let service = CsvImportService::new();
        let rows = service.parse_csv_file(temp_file.path(), '|').unwrap();
        assert_eq!(rows[0].text, "Line 1\nLine 2");
    }

    #[test]
    fn test_parse_csv_invalid_header() {
        let mut temp_file = NamedTempFile::new().unwrap();
        writeln!(temp_file, "wrong|header|format").unwrap();
        temp_file.flush().unwrap();

        let service = CsvImportService::new();
        let result = service.parse_csv_file(temp_file.path(), '|');
        assert!(matches!(result, Err(SubtitleError::InvalidCsvHeader(_, _))));
    }

    #[test]
    fn test_parse_csv_wrong_column_count() {
        let mut temp_file = NamedTempFile::new().unwrap();
        writeln!(temp_file, "start_time|end_time|text").unwrap();
        writeln!(temp_file, "00:00:01,000|00:00:02,000").unwrap(); // Missing text column
        temp_file.flush().unwrap();

        let service = CsvImportService::new();
        let result = service.parse_csv_file(temp_file.path(), '|');
        assert!(matches!(
            result,
            Err(SubtitleError::CsvParseError { line: 2, .. })
        ));
    }

    #[test]
    fn test_parse_csv_custom_delimiter() {
        let mut temp_file = NamedTempFile::new().unwrap();
        writeln!(temp_file, "start_time;end_time;text").unwrap();
        writeln!(temp_file, "00:00:01,000;00:00:02,000;First").unwrap();
        temp_file.flush().unwrap();

        let service = CsvImportService::new();
        let rows = service.parse_csv_file(temp_file.path(), ';').unwrap();
        assert_eq!(rows.len(), 1);
        assert_eq!(rows[0].text, "First");
    }

    #[test]
    fn test_parse_csv_trim_whitespace() {
        let mut temp_file = NamedTempFile::new().unwrap();
        writeln!(temp_file, "start_time|end_time|text").unwrap();
        writeln!(temp_file, "  00:00:01,000  |  00:00:02,000  |Text").unwrap();
        temp_file.flush().unwrap();

        let service = CsvImportService::new();
        let rows = service.parse_csv_file(temp_file.path(), '|').unwrap();
        assert_eq!(rows[0].start_time, "00:00:01,000");
        assert_eq!(rows[0].end_time, "00:00:02,000");
    }

    #[test]
    fn test_parse_csv_html_tags() {
        let mut temp_file = NamedTempFile::new().unwrap();
        writeln!(temp_file, "start_time|end_time|text").unwrap();
        writeln!(temp_file, "00:00:01,000|00:00:02,000|<i>Italic</i>").unwrap();
        temp_file.flush().unwrap();

        let service = CsvImportService::new();
        let rows = service.parse_csv_file(temp_file.path(), '|').unwrap();
        assert_eq!(rows[0].text, "<i>Italic</i>");
    }

    // ========== Anchored Format Tests ==========

    #[test]
    fn test_parse_anchored_single_line() {
        let mut temp_file = NamedTempFile::new().unwrap();
        writeln!(temp_file, "[1] First subtitle").unwrap();
        writeln!(temp_file, "[2] Second subtitle").unwrap();
        temp_file.flush().unwrap();

        let service = AnchoredImportService::new();
        let rows = service.parse_anchored_file(temp_file.path()).unwrap();

        assert_eq!(rows.len(), 2);
        assert_eq!(rows[0].index, 1);
        assert_eq!(rows[0].text, "First subtitle");
        assert_eq!(rows[0].line_number, 1);
        assert_eq!(rows[1].index, 2);
        assert_eq!(rows[1].text, "Second subtitle");
        assert_eq!(rows[1].line_number, 2);
    }

    #[test]
    fn test_parse_anchored_multiline() {
        let mut temp_file = NamedTempFile::new().unwrap();
        writeln!(temp_file, "[1] Line 1 of sub 1").unwrap();
        writeln!(temp_file, "Line 2 of sub 1").unwrap();
        writeln!(temp_file, "Line 3 of sub 1").unwrap();
        writeln!(temp_file, "[2] Single line").unwrap();
        temp_file.flush().unwrap();

        let service = AnchoredImportService::new();
        let rows = service.parse_anchored_file(temp_file.path()).unwrap();

        assert_eq!(rows.len(), 2);
        assert_eq!(
            rows[0].text,
            "Line 1 of sub 1\nLine 2 of sub 1\nLine 3 of sub 1"
        );
        assert_eq!(rows[1].text, "Single line");
    }

    #[test]
    fn test_parse_anchored_with_html() {
        let mut temp_file = NamedTempFile::new().unwrap();
        writeln!(temp_file, "[1] <i>Italic text</i>").unwrap();
        writeln!(temp_file, "[2] <b>Bold text</b>").unwrap();
        temp_file.flush().unwrap();

        let service = AnchoredImportService::new();
        let rows = service.parse_anchored_file(temp_file.path()).unwrap();

        assert_eq!(rows[0].text, "<i>Italic text</i>");
        assert_eq!(rows[1].text, "<b>Bold text</b>");
    }

    #[test]
    fn test_parse_anchored_invalid_no_bracket() {
        let mut temp_file = NamedTempFile::new().unwrap();
        writeln!(temp_file, "Text without bracket").unwrap();
        temp_file.flush().unwrap();

        let service = AnchoredImportService::new();
        let result = service.parse_anchored_file(temp_file.path());

        assert!(matches!(
            result,
            Err(SubtitleError::AnchoredParseError { line: 1, .. })
        ));
    }

    #[test]
    fn test_parse_anchored_empty_text() {
        let mut temp_file = NamedTempFile::new().unwrap();
        writeln!(temp_file, "[1]").unwrap();
        writeln!(temp_file, "[2] Next").unwrap();
        temp_file.flush().unwrap();

        let service = AnchoredImportService::new();
        let result = service.parse_anchored_file(temp_file.path());

        assert!(matches!(
            result,
            Err(SubtitleError::AnchoredParseError { line: 1, .. })
        ));
    }

    #[test]
    fn test_parse_anchored_empty_file() {
        let temp_file = NamedTempFile::new().unwrap();

        let service = AnchoredImportService::new();
        let result = service.parse_anchored_file(temp_file.path());

        assert!(matches!(
            result,
            Err(SubtitleError::AnchoredParseError { line: 0, .. })
        ));
    }

    #[test]
    fn test_parse_anchored_mixed_indices() {
        let mut temp_file = NamedTempFile::new().unwrap();
        writeln!(temp_file, "[5] Fifth subtitle").unwrap();
        writeln!(temp_file, "[10] Tenth subtitle").unwrap();
        writeln!(temp_file, "[1] First subtitle").unwrap();
        temp_file.flush().unwrap();

        let service = AnchoredImportService::new();
        let rows = service.parse_anchored_file(temp_file.path()).unwrap();

        assert_eq!(rows.len(), 3);
        assert_eq!(rows[0].index, 5);
        assert_eq!(rows[1].index, 10);
        assert_eq!(rows[2].index, 1);
    }

    #[test]
    fn test_parse_anchored_with_blank_lines() {
        let mut temp_file = NamedTempFile::new().unwrap();
        writeln!(temp_file, "[1] First").unwrap();
        writeln!(temp_file).unwrap(); // blank line
        writeln!(temp_file, "[2] Second").unwrap();
        temp_file.flush().unwrap();

        let service = AnchoredImportService::new();
        let rows = service.parse_anchored_file(temp_file.path()).unwrap();

        assert_eq!(rows.len(), 2);
        // Blank line becomes part of the first subtitle's text
        assert_eq!(rows[0].text, "First");
        assert_eq!(rows[1].text, "Second");
    }
}

File: pkg/importer/service.go
package importer

import (
	"encoding/csv"
	"fmt"
	"io"
	"log/slog"
	"os"
	"strconv"
	"strings"
	"unicode"

	"github.com/lebe-dev/submarine/pkg/subtitle"
)

// CsvImportService imports subtitles from CSV files. Was Rust struct
// CsvImportService.
type CsvImportService struct{}

// NewCsvImportService creates a new CsvImportService. Was CsvImportService::new.
func NewCsvImportService() *CsvImportService {
	return &CsvImportService{}
}

// validateCsvHeader validates the CSV header record. Was
// CsvImportService::validate_csv_header.
func validateCsvHeader(headers []string, delimiter rune) error {
	actual := strings.Join(headers, string(delimiter))

	if len(headers) != 3 ||
		headers[0] != "start_time" ||
		headers[1] != "end_time" ||
		headers[2] != "text" {
		return subtitle.NewInvalidCsvHeader(string(delimiter), actual)
	}

	slog.Debug("csv header validated successfully")
	return nil
}

// ParseCsvFile parses a CSV file. Was CsvImportService::parse_csv_file.
func (s *CsvImportService) ParseCsvFile(csvPath string, delimiter rune) ([]CsvSubtitleRow, error) {
	slog.Debug(fmt.Sprintf("parsing csv file: %q with delimiter '%c'", csvPath, delimiter))

	file, err := os.Open(csvPath)
	if err != nil {
		return nil, subtitle.NewIoError(err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	reader.Comma = delimiter
	// The Rust `csv` crate treats a quote as significant only at the start of a
	// field; a bare `"` in the middle of an unquoted field is kept verbatim.
	// Go's encoding/csv rejects bare quotes by default, so enable LazyQuotes to
	// match the Rust parser's tolerant quoting behavior.
	reader.LazyQuotes = true

	headers, err := reader.Read()
	if err != nil {
		return nil, subtitle.NewCsvParseError(1, fmt.Sprintf("failed to read CSV header: %s", err))
	}

	slog.Debug("validating csv header")
	if err := validateCsvHeader(headers, delimiter); err != nil {
		return nil, err
	}

	var rows []CsvSubtitleRow

	idx := 0
	for {
		record, err := reader.Read()
		if err == io.EOF {
			break
		}

		lineNumber := idx + 2 // +1 for 0-index, +1 for header

		if err != nil {
			return nil, subtitle.NewCsvParseError(lineNumber, fmt.Sprintf("failed to parse CSV record: %s", err))
		}

		if len(record) != 3 {
			return nil, subtitle.NewCsvParseError(lineNumber, fmt.Sprintf("expected 3 columns, got %d", len(record)))
		}

		slog.Debug(fmt.Sprintf("parsed csv row %d: start=%s, end=%s", lineNumber, record[0], record[1]))

		rows = append(rows, CsvSubtitleRow{
			LineNumber: lineNumber,
			StartTime:  strings.TrimSpace(record[0]),
			EndTime:    strings.TrimSpace(record[1]),
			Text:       strings.ReplaceAll(record[2], "\\n", "\n"), // Convert \n to actual newlines
		})

		idx++
	}

	slog.Debug(fmt.Sprintf("parsed %d rows from csv file", len(rows)))
	return rows, nil
}

// ParseAnchoredFile is unsupported on CsvImportService. Was
// CsvImportService::parse_anchored_file.
func (s *CsvImportService) ParseAnchoredFile(anchoredPath string) ([]AnchoredSubtitleRow, error) {
	return nil, subtitle.NewInvalidPath("CsvImportService does not support anchored format parsing")
}

// AnchoredImportService imports subtitles from anchored format files. Was Rust
// struct AnchoredImportService.
type AnchoredImportService struct{}

// NewAnchoredImportService creates a new AnchoredImportService. Was
// AnchoredImportService::new.
func NewAnchoredImportService() *AnchoredImportService {
	return &AnchoredImportService{}
}

// parseAnchoredLine parses an anchored format line to extract index and first
// line text. Returns ok=false when the line is not an [INDEX] marker. Was
// AnchoredImportService::parse_anchored_line.
func parseAnchoredLine(line string) (uint32, string, bool) {
	trimmed := strings.TrimSpace(line)
	if !strings.HasPrefix(trimmed, "[") {
		return 0, "", false
	}

	closeBracket := strings.IndexByte(trimmed, ']')
	if closeBracket < 0 {
		return 0, "", false
	}

	indexStr := trimmed[1:closeBracket]
	index, err := strconv.ParseUint(indexStr, 10, 32)
	if err != nil {
		return 0, "", false
	}

	var text string
	if closeBracket+1 < len(trimmed) {
		text = strings.TrimLeftFunc(trimmed[closeBracket+1:], unicode.IsSpace)
	} else {
		text = ""
	}

	return uint32(index), text, true
}

// ParseCsvFile is unsupported on AnchoredImportService. Was
// AnchoredImportService::parse_csv_file.
func (s *AnchoredImportService) ParseCsvFile(csvPath string, delimiter rune) ([]CsvSubtitleRow, error) {
	return nil, subtitle.NewInvalidPath("AnchoredImportService does not support CSV parsing")
}

// ParseAnchoredFile parses an anchored format file. Was
// AnchoredImportService::parse_anchored_file.
func (s *AnchoredImportService) ParseAnchoredFile(anchoredPath string) ([]AnchoredSubtitleRow, error) {
	slog.Debug(fmt.Sprintf("parsing anchored format file: %q", anchoredPath))

	content, err := os.ReadFile(anchoredPath)
	if err != nil {
		return nil, subtitle.NewIoError(err)
	}

	var rows []AnchoredSubtitleRow
	var currentIndex uint32
	hasCurrentIndex := false
	currentText := ""
	currentLineNumber := 0

	for lineNum, line := range strLines(string(content)) {
		lineNumber := lineNum + 1

		if index, firstLineText, ok := parseAnchoredLine(line); ok {
			if hasCurrentIndex {
				if strings.TrimSpace(currentText) == "" {
					return nil, subtitle.NewAnchoredParseError(currentLineNumber, fmt.Sprintf("Subtitle [%d] has empty text", currentIndex))
				}
				rows = append(rows, AnchoredSubtitleRow{
					LineNumber: currentLineNumber,
					Index:      currentIndex,
					Text:       strings.TrimSpace(currentText),
				})
			}

			currentIndex = index
			hasCurrentIndex = true
			currentText = firstLineText
			currentLineNumber = lineNumber
		} else if hasCurrentIndex {
			if currentText != "" {
				currentText += "\n"
			}
			currentText += line
		} else if strings.TrimSpace(line) != "" {
			return nil, subtitle.NewAnchoredParseError(lineNumber, "Text line found before any [INDEX] marker")
		}
	}

	if hasCurrentIndex {
		if strings.TrimSpace(currentText) == "" {
			return nil, subtitle.NewAnchoredParseError(currentLineNumber, fmt.Sprintf("Subtitle [%d] has empty text", currentIndex))
		}
		rows = append(rows, AnchoredSubtitleRow{
			LineNumber: currentLineNumber,
			Index:      currentIndex,
			Text:       strings.TrimSpace(currentText),
		})
	}

	if len(rows) == 0 {
		return nil, subtitle.NewAnchoredParseError(0, "No valid subtitle entries found in anchored file")
	}

	slog.Debug(fmt.Sprintf("parsed %d entries from anchored file", len(rows)))
	return rows, nil
}

// strLines mirrors Rust's str::lines(): splits on '\n', strips an optional
// trailing '\r' from each line, and yields no trailing empty element for a
// terminating '\n'. An empty string yields no lines.
func strLines(s string) []string {
	if s == "" {
		return nil
	}
	var lines []string
	start := 0
	for i := 0; i < len(s); i++ {
		if s[i] == '\n' {
			line := s[start:i]
			if len(line) > 0 && line[len(line)-1] == '\r' {
				line = line[:len(line)-1]
			}
			lines = append(lines, line)
			start = i + 1
		}
	}
	if start < len(s) {
		line := s[start:]
		if len(line) > 0 && line[len(line)-1] == '\r' {
			line = line[:len(line)-1]
		}
		lines = append(lines, line)
	}
	return lines
}
